url70997c83aa_Xilaew_risiko_tgz-pJ8-risiko_XMI2XMLtransformerJ8/src/risiko/main/Client2.java

```diff
@@ -0,0 +1,200 @@
+package risiko.main;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.SocketChannel;
+import java.util.Iterator;
+
+import org.eclipse.emf.common.util.EList;
+import org.eclipse.emf.ecore.EObject;
+
+import risiko.GameMonitor;
+import risiko.actions.AddPlayer;
+import risiko.actions.StartGame;
+import risiko.actions.actionFactory;
+import risiko.gamestate.Player;
+import risiko.gamestate.stateFactory;
+
+public class Client2 implements Runnable {
+	private Selector selector;
+	SocketChannel s;
+	private ByteBuffer buf = ByteBuffer.allocate(256);
+	private GameMonitor game = new GameMonitor();
+	private static final int PORT = 10523;
+	private boolean sent = false;
+	private boolean started = false;
+
+	public static void main(String[] args) {
+		SocketAddress addr = new InetSocketAddress("localhost", PORT);
+		try {
+			new Client2(addr).run();
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+	}
+
+	Client2(SocketAddress address) throws IOException {
+		this.selector = Selector.open();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code assigns a resource that might leak to a class field, causing a resource leak that slows down or crashes your system. For example, when AbstractSelectableChannel.register throws an exception, the *Client2* object isn’t constructed so you can’t close this resource afterward.

**Fix**
Ensure that if an exception is thrown, the *selector* resource is closed before the constructor is exited.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).